10-cautious/.golangci.yml

```
@@ -0,0 +1,264 @@
+        arguments:
+          - allowTypesBefore: "*testing.T"
+
+      # Basic types should not be used as a key in `context.WithValue`
+      - name: context-keys-type
+
+      # warns on some common mistakes when using defer statement.
+      - name: defer
+
+      # Importing with `.` makes the programs much harder to understand
+      - name: dot-imports
+
+      # suggest to simplify if-then-else constructions when possible
+      - name: early-return
+
+      # Empty blocks make code less readable and could be a symptom of a bug or unfinished refactoring.
+      - name: empty-block
+
+      # for better readability, variables of type `error` must be named with the prefix `err`.
+      - name: error-naming
+
+      # for better readability, the errors should be last in the list of returned values by a function.
+      - name: error-return
+
+      # for better readability, error messages should not be capitalized or end with punctuation or a newline.
+      - name: error-strings
+
+      # report when replacing `errors.New(fmt.Sprintf())` with `fmt.Errorf()` is possible
+      - name: errorf
+
+      # Checking if an error is nil to just after return the error or nil is redundant.
+      - name: if-return
+
+      # incrementing an integer variable by 1 is recommended to be done using the `++` operator
+      - name: increment-decrement
+
+      # highlights redundant else-blocks that can be eliminated from the code
+      - name: indent-error-flow
+
+      # This rule suggests a shorter way of writing ranges that do not use the second value.
+      - name: range
+
+      # receiver names in a method should reflect the struct name (p for Person, for example)
+      - name: receiver-naming
+
+      # redefining built in names (true, false, append, make) can lead to bugs very difficult to detect.
+      - name: redefines-builtin-id
+
+      # redundant else-blocks that can be eliminated from the code.
+      - name: superfluous-else
+
+      # prevent confusing name for variables when using `time` package
+      - name: time-naming
+
+      # warns when an exported function or method returns a value of an un-exported type.
+      - name: unexported-return
+
+      # spots and proposes to remove unreachable code. also helps to spot errors
+      - name: unreachable-code
+
+      # Functions or methods with unused parameters can be a symptom of an unfinished refactoring or a bug.
+      - name: unused-parameter
+
+      # warns on useless break statements in case clauses of switch and select statements
+      - name: useless-break
+
+      # report when a variable declaration can be simplified
+      - name: var-declaration
+
+      # warns when initialism, variable or package naming conventions are not followed.
+      - name: var-naming
+
+  depguard:
+    rules:
+      obsolete:
+        deny:
+          - pkg: "golang.org/x/net/context"
+            desc: "Should be replaced by standard lib context package (Go 1.7+)"
```

I'm wondering if it would be clearer to do

```suggestion
          - pkg: "golang.org/x/net/context"
            desc: "Should be replaced by standard lib "context" package (Go 1.7+)"
```

---

[`staticcheck`](https://staticcheck.dev/docs/checks/#SA1019) for example formats it like that:
```
staticcheck: SA1019: "io/ioutil" has been deprecated since Go 1.19: As of Go 1.16, the same functionality is now provided by pack
age [io] or package [os], and those implementations should be preferred in new code. See the specific function documentation for details.
```

---

This also yield the question of possible redundancy between this and `staticcheck`. It might be worth enabling the `staticcheck` rule `SA1019` and seeing what's missing. IMHO contributing to this rule instead of maintaining a custom list of deprecated pkg is better.

---

Thanks for your valuable feedbacks here.

I will indeed have to validate the std lib ones against staticcheck

The problem I have is staticcheck and its maintainer is that the project is barely maintained and maintainer refuses PR

---

Ah, I didn't know that. Then it's probably good to have a separate check :)

---

here

https://github.com/dominikh/.github/blob/main/CONTRIBUTING.md#pull-requests

I dislike the "I don't have time to review your bad code. I prefer to code everything by my own" tone

---

Lol, indeed.

---

Yes. I did. I even filled an issue on the exptostd repository last week. It was about something that led me to write this here.

I had also opened one on usestdlibvar about syscall constants.

This PR needs some reworks. I thought about mentioning exptostd and keep a comment in the file

Maybe it's simply because exptostd was not yet available via golangci-lint when I started working on these changes a few weeks ago

---

Yes. I did. I even filled an issue on the exptostd repository last week. It was about something that led me to write this here.

I had also opened one on usestdlibvar about syscall constants.

This PR needs some reworks. I thought about mentioning exptostd and keep a comment in the file

Maybe it's simply because exptostd was not yet available via golangci-lint when I started working on these changes a few weeks ago